tsdb/agent/db.go

```diff
@@ -646,56 +660,67 @@ func (db *DB) Close() error {
 type appender struct {
 	*DB
 
-	pendingSeries  []record.RefSeries
-	pendingSamples []record.RefSample
+	pendingSeries    []record.RefSeries
+	pendingSamples   []record.RefSample
+	pendingExamplars []record.RefExemplar
 }
 
 func (a *appender) Append(ref uint64, l labels.Labels, t int64, v float64) (uint64, error) {
-	if ref == 0 {
-		return a.Add(l, t, v)
-	}
-	return ref, a.AddFast(ref, t, v)
-}
+	series := a.series.GetByID(ref)
+	if series == nil {
+		// Ensure no empty or duplicate labels have gotten through. This mirrors the
+		// equivalent validation code in the TSDB's headAppender.
+		l = l.WithoutEmpty()
+		if len(l) == 0 {
+			return 0, errors.Wrap(tsdb.ErrInvalidSample, "empty labelset")
+		}
 
-func (a *appender) Add(l labels.Labels, t int64, v float64) (uint64, error) {
-	hash := l.Hash()
-	series := a.series.GetByHash(hash, l)
-	if series != nil {
-		return series.ref, a.AddFast(series.ref, t, v)
-	}
+		if lbl, dup := l.HasDuplicateLabelNames(); dup {
+			return 0, errors.Wrap(tsdb.ErrInvalidSample, fmt.Sprintf(`label name "%s" is not unique`, lbl))
+		}
 
-	// Ensure no empty or duplicate labels have gotten through. This mirrors the
-	// equivalent validation code in the TSDB's headAppender.
-	l = l.WithoutEmpty()
-	if len(l) == 0 {
-		return 0, errors.Wrap(tsdb.ErrInvalidSample, "empty labelset")
-	}
+		var created bool
+		series, created = a.getOrCreate(l)
+		if created {
+			a.pendingSeries = append(a.pendingSeries, record.RefSeries{
+				Ref:    series.ref,
+				Labels: l,
+			})
 
-	if lbl, dup := l.HasDuplicateLabelNames(); dup {
-		return 0, errors.Wrap(tsdb.ErrInvalidSample, fmt.Sprintf(`label name "%s" is not unique`, lbl))
+			a.metrics.numActiveSeries.Inc()
+		}
 	}
 
-	ref := a.nextRef.Inc()
-	series = &memSeries{ref: ref, lset: l, lastTs: t}
+	series.Lock()
+	defer series.Unlock()
+
+	// Update last recorded timestamp. Used by Storage.gc to determine if a
+	// series is stale.
+	series.lastTs = t
 
-	a.pendingSeries = append(a.pendingSeries, record.RefSeries{
-		Ref:    ref,
-		Labels: l,
-	})
 	a.pendingSamples = append(a.pendingSamples, record.RefSample{
-		Ref: ref,
+		Ref: series.ref,
 		T:   t,
 		V:   v,
 	})
 
-	a.series.Set(hash, series)
-
-	a.metrics.numActiveSeries.Inc()
 	a.metrics.totalAppendedSamples.Inc()
-
 	return series.ref, nil
 }
 
+func (a *appender) getOrCreate(l labels.Labels) (series *memSeries, created bool) {
+	hash := l.Hash()
+
+	series = a.series.GetByHash(hash, l)
+	if series != nil {
+		return series, false
+	}
+
+	series = &memSeries{ref: a.nextRef.Inc(), lset: l}
+	a.series.Set(l.Hash(), series)
```

Comment:
Hash is expensive, so we should be re-using the hash from the above. Or use the same `getOrCreate(hash uint64, lset labels.Labels)` signature from TSDB which mandates reuse of the hash.